packages/notifier/src/asyncIterableAdaptor.js

```diff
@@ -0,0 +1,123 @@
+      let localUpdateCount;
+      let myIterationResultP;
+      return harden({
+        next: () => {
+          if (!myIterationResultP) {
+            // In this adaptor, once `next()` is called and returns an
+            // unresolved promise, `myIterationResultP`, and until
+            // `myIterationResultP` is fulfilled with an
+            // iteration result, further `next()` calls will return the same
+            // `myIterationResultP` promise again without asking the notifier
+            // for more updates. If there's already an unanswered ask in the
+            // air, all further asks should just reuse the result of that one.
+            //
+            // This reuse behavior is only needed for code that uses the async
+            // iterator protocol explicitly. When this async iterator is
+            // consumed by a for/await/of loop, `next()` will only be called
+            // after the promise for the previous iteration result has
+            // fulfilled. If it fulfills with `done: true`, the for/await/of
+            // loop will never call `next()` again.
+            //
+            // See
+            // https://2ality.com/2016/10/asynchronous-iteration.html#queuing-next()-invocations
+            // for an explicit use that sends `next()` without waiting.
+            myIterationResultP = E(notifierP)
+              .getUpdateSince(localUpdateCount)
+              .then(({ value, updateCount }) => {
+                localUpdateCount = updateCount;
+                const done = localUpdateCount === undefined;
+                if (!done) {
+                  // Once the outstanding question has been answered, stop
+                  // using that answer, so any further `next()` questions
+                  // cause a new `getUpdateSince` request.
+                  //
+                  // But only if more answers are expected. Once the notifier
+                  // is `done`, that was the last answer so reuse it forever.
+                  myIterationResultP = undefined;
+                }
+                return harden({ value, done });
+              });
+          }
+          return myIterationResultP;
+        },
+      });
+    },
+  });
+};
+
+/**
+ * This reads from `iteratorP` updating `updater` with each successive value.
+ * `iteratorP` can be a sync or async iterator, or a promise for a local or
+ * remote iterator, or a presence of a remote iterator. When `iteratorP` is
+ * done, `updater` is resoved, so that the corresponding notifier is
+ * also done.
+ *
+ * The `updater` must be local and is called synchronously, so this adapter
+ * must be co-located with the updater.
+ */
+const updateFromIterator = (updater, iteratorP) => {
+  return E(iteratorP)
+    .next()
+    .then(
+      ({ value, done }) => {
+        if (done) {
+          updater.finish(value);
+          return undefined;
+        }
+        updater.updateState(value);
+        return updateFromIterator(updater, iteratorP);
```

Comment:
I'm not sure if there needs to be any special handling here.  I found when I write recursive functions with promises, they are not as well-behaved as an iterative `await` within a `for` loop.

At least on Node.js, a recursive function as follows will consume all heap:
```js
function loop1() {
  return Promise.resolve().then(loop1);
}
```

where the iterative:

```js
async function loop2() {
  for (;;) {
    await Promise.resolve();
  }
}
```
does not (though it eventually fails for a different reason).

```
RangeError: Value undefined out of range for undefined options property undefined
    at Map.set (<anonymous>)
    at AsyncHook.init (domain.js:68:15)
    at PromiseWrap.emitInitNative (internal/async_hooks.js:127:43)
    at loop2 (repl:1:51)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

Not sure what the situation is on XS.

Reply:
What does that `RangeError` mean? I have no idea.

Reply:
It's a bug in Node, specifically (not v8).

Reply:
If I run loop1 and loop2 in a Brave tab, they both consume 100% CPU, and loop1's heap grows quickly, but loop2's heap does not grow.

Reply:
I kept the recursion, but avoided the promise chain causing the memory leak. PTAL.